destination/clickhouse.go

```diff
@@ -30,146 +31,185 @@ type ClickHouseConnection struct {
 }
 
 func GetClickHouseConnection(ctx context.Context, configuration map[string]string) (*ClickHouseConnection, error) {
-	hostname := fmt.Sprintf("%s:%s",
-		GetWithDefault(configuration, "hostname", "localhost"),
-		GetWithDefault(configuration, "port", "9000"))
-	username := GetWithDefault(configuration, "username", "default")
-	password := GetWithDefault(configuration, "password", "")
-	database := GetWithDefault(configuration, "database", "default")
-	options := &clickhouse.Options{
-		Addr: []string{hostname},
-		Auth: clickhouse.Auth{
-			Username: username,
-			Password: password,
-			Database: database,
-		},
-		Protocol: clickhouse.Native,
-		Settings: clickhouse.Settings{
-			"allow_experimental_object_type": 1,
-			"date_time_input_format":         "best_effort",
-			//"async_insert":                   1,
-			//"wait_for_async_insert":          1,
-		},
-		ClientInfo: clickhouse.ClientInfo{
-			Products: []struct {
-				Name    string
-				Version string
-			}{
-				{Name: "fivetran-destination", Version: Version},
+	return RetryNetErrorWithData(func() (*ClickHouseConnection, error) {
+		hostname := fmt.Sprintf("%s:%s",
+			GetWithDefault(configuration, "hostname", "localhost"),
+			GetWithDefault(configuration, "port", "9000"))
+		username := GetWithDefault(configuration, "username", "default")
+		password := GetWithDefault(configuration, "password", "")
+		database := GetWithDefault(configuration, "database", "default")
+		options := &clickhouse.Options{
+			Addr: []string{hostname},
+			Auth: clickhouse.Auth{
+				Username: username,
+				Password: password,
+				Database: database,
 			},
-		},
-	}
-	ssl := GetWithDefault(configuration, "ssl", "false")
-	if ssl == "true" {
-		options.TLS = &tls.Config{InsecureSkipVerify: true}
-	}
-	conn, err := clickhouse.Open(options)
-	if err != nil {
-		return nil, err
-	}
-	return &ClickHouseConnection{database, ctx, conn}, nil
+			Protocol: clickhouse.Native,
+			Settings: clickhouse.Settings{
+				"allow_experimental_object_type": 1,
+				"date_time_input_format":         "best_effort",
+			},
+			MaxOpenConns: int(*maxOpenConnections),
+			MaxIdleConns: int(*maxIdleConnections),
+			ClientInfo: clickhouse.ClientInfo{
+				Products: []struct {
+					Name    string
+					Version string
+				}{
+					{Name: "fivetran-destination", Version: Version},
+				},
+			},
+		}
+		ssl := GetWithDefault(configuration, "ssl", "false")
+		if ssl == "true" {
+			options.TLS = &tls.Config{InsecureSkipVerify: true}
```

Comment:
any reason we force `InsecureSkipVerify: true`? It's fine to have it configurable by user, but defaulting it to `true` is risky.

Reply:
I agree, that's a rudiment that I initially took from this example: https://clickhouse.com/docs/en/integrations/go#copy-in-some-sample-code

Reply:
Fixed.